Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -150,41 +158,78 @@ namespace ROS2
+            builder.AddField("rgba", sensor_msgs::msg::PointField::UINT32);
+        }
+
+        sensor_msgs::msg::PointCloud2 message = builder.Get();
 
-        sensor_msgs::PointCloud2Iterator<float> messageXIt(message, "x");
-        sensor_msgs::PointCloud2Iterator<float> messageYIt(message, "y");
-        sensor_msgs::PointCloud2Iterator<float> messageZIt(message, "z");
+        Pc2MsgIt<float> messageXIt(message, "x");
+        Pc2MsgIt<float> messageYIt(message, "y");
+        Pc2MsgIt<float> messageZIt(message, "z");
 
         const auto positionField = results.GetConstFieldSpan<RaycastResultFlags::Point>().value();
         auto positionIt = positionField.begin();
 
-        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> messageIntensityIt;
+        AZStd::optional<Pc2MsgIt<float>> messageIntensityIt;
         AZStd::optional<RaycastResults::FieldSpan<RaycastResultFlags::Intensity>::const_iterator> intensityIt;
-        if (isIntensityEnabled)
+        if (results.IsFieldPresent<RaycastResultFlags::Intensity>())
         {
-            messageIntensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+            messageIntensityIt = Pc2MsgIt<float>(message, "intensity");
             intensityIt = results.GetConstFieldSpan<RaycastResultFlags::Intensity>().value().begin();
         }
 
+        struct MessageSegmentationIterators
+        {
+            Pc2MsgIt<int32_t> m_entityIdIt;
+            Pc2MsgIt<uint8_t> m_classIdIt;
+            Pc2MsgIt<uint32_t> m_rgbaIt;
+        };
+
+        AZStd::optional<MessageSegmentationIterators> messageSegDataIts;
+        AZStd::optional<RaycastResults::FieldSpan<RaycastResultFlags::SegmentationData>::const_iterator> segDataIt;
+        if (results.IsFieldPresent<RaycastResultFlags::SegmentationData>())
+        {
+            messageSegDataIts = MessageSegmentationIterators{
+                Pc2MsgIt<int32_t>(message, "entity_id"),
+                Pc2MsgIt<uint8_t>(message, "class_id"),
+                Pc2MsgIt<uint32_t>(message, "rgba"),
+            };
+
+            segDataIt = results.GetConstFieldSpan<RaycastResultFlags::SegmentationData>().value().begin();
+        }
+
         const auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>();
         const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
+
+        auto* classSegmentationInterface = ClassSegmentationInterface::Get();
+        AZ_Warning(
+            __func__,
+            !results.IsFieldPresent<RaycastResultFlags::SegmentationData>() || classSegmentationInterface,
+            "Segmentation data was requested but the Class Segmentation interface was unavailable. Unable to fetch segmentation class "
+            "data. Please make sure to add the Class Segmentation Configuration Component to the Level Entity for this feature to work "
+            "properly.");
+
         for (size_t i = 0; i < results.GetCount(); ++i)
```

Comment:
This probably needs threading, as it will probably take quite a bit of time if run with 10m points

Reply:
I would move this to a separate pr. I would prefer to run some profiling and find out if we don't have any different bottlenecks. Besides, due to the fact that the processed data is in a few contiguous blocks of memory (one per each field) this also runs quite fast due to cache friendliness. Again, at this point I can only speculate as I've yet to profile this code.